### PR TITLE
Mandos new NFT syntax

### PR DESCRIPTION
**Pull request by andrei-marinica**

### CODE REVIEW

**Review by sasurobert — APPROVED** [review]

File: arwenmandos/stepCheckAccounts.go [hunk 1]
```diff
@@ -239,16 +242,17 @@ func checkTokenState(tokenName string, expectedTokenData *mj.CheckESDTData, acco
...
-		return fmt.Errorf("bad ESDT balance. Token %s: Want: %d. Have: %d",
-			tokenName, expectedTokenData.Value.Value, accountTokenData.Value)
-	}
-
-	metadataFromBytes := builtInFunctions.ESDTUserMetadataFromBytes(accountTokenData.Properties)
-	if !expectedTokenData.Frozen.CheckBool(metadataFromBytes.Frozen) {
-		return fmt.Errorf("bad ESDT frozen flag. Token %s: Want: %t. Have: %t",
-			tokenName, expectedTokenData.Frozen.Value > 0, metadataFromBytes.Frozen)
-	}
+	// TODO: fix mapping
```
Comment (sasurobert) [inline arwenmandos/stepCheckAccounts.go:245]: Maybe fix this before merging into master

**Review by sasurobert — CHANGES_REQUESTED** [review]

File: mock/world/mockESDTAccount.go [hunk 2]
```diff
@@ -238,3 +242,70 @@ func (a *Account) GetTokenNames() ([][]byte, error) {
...
+			LastNonce:       0,
+			Roles:           nil,
+		}
+		resultMap[tokenName] = resultObj
+	}
+	return resultObj
+}
+
+// GetFullMockESDTData returns the information about all the ESDT tokens held by the account.
+func (a *Account) GetFullMockESDTData() (map[string]*MockESDTData, error) {
```
Comment (camilbancioiu) [inline mock/world/mockESDTAccount.go:269]: Good, but this needs to be broken down into smaller functions.

File: mock/world/mockESDTAccount.go [hunk 3]
```diff
@@ -238,3 +242,70 @@ func (a *Account) GetTokenNames() ([][]byte, error) {
...
+
+// MockESDTData groups together all instances of a token (same token name, different nonces).
+type MockESDTData struct {
+	TokenIdentifier []byte
+	Instances       []*esdt.ESDigitalToken
+	LastNonce       uint64
+	Roles           [][]byte
+}
+
+func getOrCreateMockESDTData(tokenName string, resultMap map[string]*MockESDTData) *MockESDTData {
```
Comment (camilbancioiu) [inline mock/world/mockESDTAccount.go:254]: Optionally move this function below `GetFullMockESDTData()`.

File: mock/world/mockESDTAccount.go [hunk 4]
```diff
@@ -238,3 +242,70 @@ func (a *Account) GetTokenNames() ([][]byte, error) {
...
+		resultMap[tokenName] = resultObj
+	}
+	return resultObj
+}
+
+// GetFullMockESDTData returns the information about all the ESDT tokens held by the account.
+func (a *Account) GetFullMockESDTData() (map[string]*MockESDTData, error) {
+	resultMap := make(map[string]*MockESDTData)
+	for key := range a.Storage {
+		if IsTokenKey([]byte(key)) {
```
Comment (camilbancioiu) [inline mock/world/mockESDTAccount.go:272]: Optionally declare `key` as `[]bytes` once, before the `if-else` block.
```suggestion
	for strKey := range a.Storage {
	        key := []byte(strKey)
		if IsTokenKey(key) {
```

File: mock/world/mockESDTAccount.go [hunk 5]
```diff
@@ -238,3 +242,70 @@ func (a *Account) GetTokenNames() ([][]byte, error) {
...
+					tokenNameFromKey := GetTokenNameFromKey([]byte(key))
+					tokenInstance.TokenMetaData = &esdt.MetaData{
+						Name:  tokenNameFromKey,
+						Nonce: 0,
+					}
+					tokenName = string(tokenNameFromKey)
+				} else {
+					// the key also contains the nonce, we take the token identifier from the metadata
+					tokenName = string(tokenInstance.TokenMetaData.Name)
+				}
```
Comment (camilbancioiu) [inline mock/world/mockESDTAccount.go:291]: Replace with a function that takes `tokenInstance` and `key` and returns `tokenName`.

File: arwenmandos/stepCheckAccounts.go [hunk 6]
```diff
@@ -171,86 +193,131 @@ func (ae *ArwenTestExecutor) checkAccountESDT(expectedAcct *mj.CheckAccount, mat
...
+			if !expectedInstance.Balance.Check(accountInstance.Value) {
+				errors = append(errors, fmt.Errorf("bad ESDT balance. Account: %s. Token: %s. Nonce: %d. Want: %s. Have: %d",
+					accountAddress,
+					tokenName,
+					nonce,
+					expectedInstance.Balance.Original,
+					accountInstance.Value))
+			}
+
+			// TODO: check metadata/properties
```
Comment (camilbancioiu) [inline arwenmandos/stepCheckAccounts.go:278]: Is this TODO important?

**Review by camilbancioiu — CHANGES_REQUESTED** [review]

**Comment by camilbancioiu** [comment]
Some refactoring needed.

File: arwenmandos/util.go [hunk 7]
```diff
@@ -45,43 +45,30 @@ func convertAccount(testAcct *mj.Account) (*worldmock.Account, error) {
...
-				Name:  tokenName,
-				Nonce: tokenNonce,
-			},
+		for _, instance := range mandosESDTData.Instances {
+			tokenNonce := instance.Nonce.Value
+			tokenKey := worldmock.MakeTokenKey(tokenName, tokenNonce)
+			tokenBalance := instance.Balance.Value
+			tokenData := &esdt.ESDigitalToken{
+				Value:      tokenBalance,
+				Type:       uint32(core.Fungible),
```
Comment (dorin-iancu) [inline arwenmandos/util.go:55]: Why is type always Fungible? Why have non-zero nonce for Fungible ESDT?

File: mandos-go/json/parse/parseESDTTx.go [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+	if !isMap {
+		return nil, errors.New("unmarshalled account object is not a map")
+	}
+
+	esdtData := mj.ESDTTxData{}
+	var err error
+
+	for _, kvp := range fieldMap.OrderedKV {
+		switch kvp.Key {
+		case "tokenIdentifier":
```
Comment (dorin-iancu) [inline mandos-go/json/parse/parseESDTTx.go:22]: Optional: rename this field to "tokenId" in mandos.

File: mandos-go/json/write/writeAccount.go [hunk 9]
```diff
@@ -5,7 +5,8 @@ import (
 	oj "github.com/ElrondNetwork/arwen-wasm-vm/mandos-go/orderedjson"
 )
 
-func accountsToOJ(accounts []*mj.Account) oj.OJsonObject {
+// AccountsToOJ converts a mandos-format account to an ordered JSON representation.
+func AccountsToOJ(accounts []*mj.Account) oj.OJsonObject {
```
Comment (dorin-iancu) [inline mandos-go/json/write/writeAccount.go:9]: Why uppercase "A"? Isn't the standard to have all functions starting with lowercase letter in golang?

Comment (andrei-marinica) [reply mandos-go/json/write/writeAccount.go:9]: `UpperCamelCase` = public
`lowerCamelCase` = private in Go

File: arwenmandos/stepCheckAccounts.go (hunk 6 above)
Comment (andrei-marinica) [reply arwenmandos/stepCheckAccounts.go:278]: yes, coming in the next PR

File: arwenmandos/util.go (hunk 7 above)
Comment (andrei-marinica) [reply arwenmandos/util.go:55]: Not yet properly implemented. Will do

Comment (andrei-marinica) [reply arwenmandos/util.go:55]: ... in the next PR

File: mandos-go/expression/reconstructor/reconstructor.go [hunk 10]
```diff
@@ -55,12 +58,19 @@ func unknownByteArrayPretty(bytes []byte) string {
...
 	}
 
+	// fully interpret as string
 	if canInterpretAsString(bytes) {
-		return fmt.Sprintf("0x%s (``%s)", hex.EncodeToString(bytes), string(bytes))
+		return fmt.Sprintf("0x%s (str:%s)", hex.EncodeToString(bytes), string(bytes))
+	}
+
+	// interpret as number
+	if len(bytes) < maxBytesInterpretedAsNumber {
```
Comment (camilbancioiu) [inline mandos-go/expression/reconstructor/reconstructor.go:67]: Shouldn't the interpretation as number have priority over the interpretation as string?

**Review by camilbancioiu — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]

**Pull request merged**